HOTT-1107: Adds date page

willfish opened the pull request:
### Jira link

https://transformuk.atlassian.net/browse/HOTT-1107

### What?

**The change dates page defaults to todays date when no query date is supplied**

![Screenshot from 2021-11-25 09-45-32](https://user-images.githubusercontent.com/8156884/143419116-5e2c3e79-bb0d-4bff-bc67-401e024cd754.png)

**Basic date validations feed into the form correctly**
![Screenshot from 2021-11-25 09-45-49](https://user-images.githubusercontent.com/8156884/143419121-2f927ca3-17b8-47b4-9c4d-86bdd695d773.png)

**Query date populates the form input values**
![Screenshot from 2021-11-25 09-46-17](https://user-images.githubusercontent.com/8156884/143419126-69fbb972-369f-40de-8e30-1354bb97c23d.png)

**Commodity context now with link**
![Screenshot from 2021-11-25 09-46-53](https://user-images.githubusercontent.com/8156884/143419133-a5be5a47-20cb-41f8-acae-e33ac4b326b5.png)

**Declarable heading context now with link**
![Screenshot from 2021-11-25 09-47-15](https://user-images.githubusercontent.com/8156884/143419140-c890d93b-2750-4b58-ae2a-13320823e6b8.png)

**Non-declarable heading context now with link**
![Screenshot from 2021-11-25 09-47-32](https://user-images.githubusercontent.com/8156884/143419142-0d2b7bd8-3254-4054-9154-6e1b97336dcc.png)

**Chapter context now with link**
![Screenshot from 2021-11-25 09-47-46](https://user-images.githubusercontent.com/8156884/143419146-e6e40249-ba42-4870-88e1-f2a8e86fa2d5.png)

**Section has no context menu and cannot change the date**
![Screenshot from 2021-11-25 09-48-00](https://user-images.githubusercontent.com/8156884/143419151-78ecbd49-a62f-448a-a489-04ea15a2daed.png)

I have added/removed/altered:

- [x] Adds change date controller, model, routes and views to enable making changes to the import date
- [x] Migrates breadcrumbs to use home_path
- [x] Adds setting of the goods nomenclature item id in chapter page
- [x] Removes now-unused date forms/partials and their references
- [x] Adds specs covering all of the
- [x] Adds links in context summary lists to change the date
- [x] Lints some touched files

### Why?

I am doing this because:

- This is a feature request and is required to do dates in a compliant way

alexdesi commented on app/helpers/declarable_helper.rb:
```diff
@@ -40,23 +40,28 @@ def declarable_stw_link(declarable, search)
+      heading_path(goods_nomenclature_path_opts)
     else
-      commodity_path(declarable_path_opts)
+      commodity_path(goods_nomenclature_path_opts)
```
it looks the path can be a method inside
Chapter, Heading and commodity
and we'd not need the case

alexdesi commented on app/views/dates/show.html.erb:
```diff
@@ -0,0 +1,98 @@
+
+      <div class="govuk-button-group">
+        <%= submit_tag 'Update date', class: 'govuk-button', name: nil %>
+
```
I am not wrong, Madhu said that this is not working

willfish committed:
Adds date page with dynamic redirects for active goods nomenclature

Also:

- adds change date model, controller, routes and specs
- lints some touched files
- migrates breadcrumbs to use home_path
- adds setting of the goods nomenclature item id in chapter page
- removes now-unused date forms/partials
- adds specs covering all of the
- adds links in context summary lists to change the date

alexdesi commented on app/controllers/change_dates_controller.rb:
```diff
@@ -0,0 +1,59 @@
+class ChangeDatesController < ApplicationController
```
just calling it `DatesController`  sounds better,
but maybe there is a reason to call it `ChangeDatesController`

willfish replied:
I just picked what Jeremy suggested offline. I originally had DatesController

willfish replied:
I really don't have strong opinions

willfish replied:
So semantically ImportDate would make more sense

alexdesi commented on app/models/change_date.rb:
```diff
@@ -0,0 +1,49 @@
+
+  def year
+    import_date&.year
+  end
```
maybe you want to delegate these 3 methods to  import date using allow_nil?

willfish replied:
Good idea :)

alexdesi reviewed (commented):
Just a couple of comments.

willfish committed:
Revert home path

willfish committed:
Code review

alexdesi commented on app/models/change_date.rb:
```diff
@@ -0,0 +1,39 @@
+
+  attribute :import_date, :date
+
+  delegate :day, :month, :year, to: :import_date, allow_nil: true
```
nice 👍

willfish committed:
ChangeDate -> ImportExportDate

alexdesi commented on app/controllers/import_export_dates_controller.rb:
```diff
@@ -0,0 +1,59 @@
+  end
+
+  def show_import_export_date_params
+    import_export_date_params = if params[:day] && params[:month] && params[:year]
+                                  {
+                                    'import_date(3i)': params[:day],
+                                    'import_date(2i)': params[:month],
+                                    'import_date(1i)': params[:year],
+                                  }
+                                else
+                                  {
+                                    'import_date(3i)': today.day.to_s,
+                                    'import_date(2i)': today.month.to_s,
+                                    'import_date(1i)': today.year.to_s,
+                                  }
+                                end
```
this is kind of equivalent:
```
import_export_date_params = {
                                    'import_date(3i)': params[:day] || today.day.to_s,
                                    'import_date(2i)': params[:month] || today.month.to_s,
                                    'import_date(1i)': params[:year] || params[:year],
                                  }
```

willfish replied:
I need all three of the query params to be set sadly - I had thought of doing something similar
